share/CMakeJS.cmake

```diff
@@ -0,0 +1,640 @@
+    string(REGEX REPLACE "[\r\n\"]" "" CMAKE_JS_LIB "${CMAKE_JS_LIB}")
+
+    set(CMAKE_JS_INC "${CMAKE_JS_INC}" CACHE STRING "cmake-js include directory." FORCE)
+    set(CMAKE_JS_SRC "${CMAKE_JS_SRC}" CACHE STRING "cmake-js source file." FORCE)
+    set(CMAKE_JS_LIB "${CMAKE_JS_LIB}" CACHE STRING "cmake-js lib file." FORCE)
+
+    # TODO: At this point, some warnings may occur re: the below (still investigating);
+    # Define either NAPI_CPP_EXCEPTIONS or NAPI_DISABLE_CPP_EXCEPTIONS.
+    #set (NAPI_CPP_EXCEPTIONS TRUE CACHE STRING "Define either NAPI_CPP_EXCEPTIONS or NAPI_DISABLE_CPP_EXCEPTIONS")
+    add_definitions(-DNAPI_CPP_EXCEPTIONS) # Also needs /EHsc
+    # add_definitions(-DNAPI_DISABLE_CPP_EXCEPTIONS)
+
+else ()
+
+    # ... we already are calling via npm/cmake-js, so we should already have all the vars we need!
+    if(VERBOSE)
+        message(DEBUG "CMakeJS Calling...")
+    endif()
+
+endif ()
+
+unset(CMAKE_JS_INC_FILES) # prevent repetitive globbing on each run
+file(GLOB CMAKE_JS_INC_FILES "${CMAKE_JS_INC}/*.h")
+file(GLOB CMAKE_JS_INC_FILES "${CMAKE_JS_INC}/**/*.h")
+source_group("cmake-js v${_version} Node ${NODE_VERSION}" FILES "${CMAKE_JS_INC_FILES}")
+
+# Log the vars to the console for sanity...
+if(VERBOSE)
+    message(DEBUG "CMAKE_JS_INC = ${CMAKE_JS_INC}")
+    message(DEBUG "CMAKE_JS_SRC = ${CMAKE_JS_SRC}")
+    message(DEBUG "CMAKE_JS_LIB = ${CMAKE_JS_LIB}")
+endif()
+
+#[=============================================================================[
+Get the in-use NodeJS binary for executing NodeJS commands in CMake scripts.
+
+Provides
+
+::
+
+  NODE_EXECUTABLE, the NodeJS runtime binary being used
+  NODE_VERSION, the version of the NodeJS runtime binary being used
+
+]=============================================================================]#
+function(cmakejs_acquire_node_executable)
+    find_program(NODE_EXECUTABLE
+      NAMES "node" "node.exe"
+      PATHS "$ENV{PATH}" "$ENV{ProgramFiles}/nodejs"
+      DOC "NodeJs executable binary"
+      REQUIRED
+    )
+    if (NOT NODE_EXECUTABLE)
+        message(FATAL_ERROR "NodeJS installation not found! Please check your paths and try again.")
+        return()
+    endif()
+
+    execute_process(
+      COMMAND "${NODE_EXECUTABLE}" "--version"
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+      OUTPUT_VARIABLE NODE_VERSION
+    )
+    string(REGEX REPLACE "[\r\n\"]" "" NODE_VERSION "${NODE_VERSION}")
+    set(NODE_VERSION "${NODE_VERSION}" CACHE STRING "" FORCE)
+
+    if(VERBOSE)
+        message(STATUS "NODE_EXECUTABLE: ${NODE_EXECUTABLE}")
+        message(STATUS "NODE_VERSION: ${NODE_VERSION}")
+    endif()
+endfunction()
+
+if(NOT DEFINED NODE_EXECUTABLE)
+    cmakejs_acquire_node_executable()
+    message(STATUS "NODE_EXECUTABLE: ${NODE_EXECUTABLE}")
+    message(STATUS "NODE_VERSION: ${NODE_VERSION}")
+endif()
+
+# Resolve NodeJS development headers
+# TODO: This code block is quite problematic, since:
```

Comment:
Yeah, I don't think this should be done. In my mind, if `node_modules` is missing that means that `cmake-js` and therefore this cmake file will not be available and so is a non-issue, or for safety it should simply fail here

Reply:
Fair - if the logic holds up with the codeblock removed, better than adding complexity with the ```add_custom_target()``` idea.

I'll let you know if any issues do arise in my tests, but I see what you mean!

Reply:
In deeper reflection (sorry not at my machine rn):

That codeblock is the bit that makes it all work *without* running from cmake-js.

Users can build everything with just normal CMake CLI commands - it works because of that hacky line. I obviously feel the solution needs a lot of work, but it does build without cmake-js!

Check my [demo commands](https://github.com/nathanjhood/NapiAddon/tree/main/package.json).

Reply:
I think this is ultimately going to come down to how you install cmake-js, its not necessary to be able to use plain cmake

If you install it globally on your machine, then this block might be useful but as you point out risks performing a build as part of the `yarn install`.

If you install it as a dependency of the project, then this block is not needed as this file has already been loaded from the node_modules so it must exist.

Ideally we should support both cases, so the question becomes how much to do automatically (in the first iteration).  
I personally would be ok with simply failing saying that the directory doesnt exist. This is no different to other cmake projects which rely on system dependencies to be installed.  

And if someone feels strongly opposed, then this could easily be added later on. Perhaps the package manager problem can be solved by reading it from a variable? I don't know how to solve it might trigger another build, my first thought was using an environment variable which would tell cmake-js to skip the build but that would cause issues if there was another cmake-js based dependency to be installed.


-----

Side note, I have never been a fan of installing things like this globally. Wouldn't everyone end up having issues once different projects require different versions of whatever you installed globally?

Reply:
> Side note, I have never been a fan of installing things like this globally. Wouldn't everyone end up having issues once different projects require different versions of whatever you installed globally?

Globally installing with cmake-js "just works" for me, and usually fixes any missing headers etc that I've ever encountered. Gladly, it works perfectly in sync with ```nvm``` - so, my cmake-js libnode-dev filetree version is in sync with my runtime node version, without any effort from my side. It does what it does well, but one slight issue with this is that I don't necessarily want a '.cmakejs' dir in my home folder, containing multiple filetrees. Would be nicer if these could all be kept project-local, either in 'node_modules' or the build dir, or anywhere where we can easily just blow them away when closing down the project.

But yeah, nvm and cmake-js always worked well here!

Reply:
> It does what it does well, but one slight issue with this is that I don't necessarily want a '.cmakejs' dir in my home folder, containing multiple filetrees.

I believe that the `.cmakejs` folder contains a cache of the 'full' headers that have been downloaded, to avoid needing to constantly redownload them. If they are only cached inside of the project folder, then there will be a lot more cache misses. (when switching project, when doing a new clone, when cleaning out all uncommitted files)